### ISSUE TITLE
Review: Vision Transformer (ViT).

### ISSUE THREAD

**Issue by NorbertZheng** [NorbertZheng/read-papers#51]
Sik-Ho Tsang. [Review: Vision Transformer (ViT)](https://sh-tsang.medium.com/review-vision-transformer-vit-406568603de0).
Dosovitskiy A, Beyer L, Kolesnikov A, et al. [An image is worth 16x16 words: Transformers for image recognition at scale](https://arxiv.org/pdf/2010.11929.pdf).

**Comment by NorbertZheng** [comment]
## Overview
 - [Transformer](https://sh-tsang.medium.com/review-attention-is-all-you-need-transformer-96c787ecdec1) architecture has become the de-facto standard for natural language processing tasks.
 - Vision Transformer (ViT), a pure [Transformer](https://sh-tsang.medium.com/review-attention-is-all-you-need-transformer-96c787ecdec1) applied directly to sequences of image patches can perform very well on image classification tasks, where CNN is not necessary.

**Comment by NorbertZheng** [comment]
## Vision Transformer (ViT)
Vision Transformer (ViT) Network Architecture.
![image](https://user-images.githubusercontent.com/37235023/216051330-788c7328-1064-4dcd-a603-18f8f06aabf8.png)

To handle 2D images, the image $x$ is reshaped from $H\times W\times C$ into a sequence of flattened 2D patches $x_{p}$, with the shape of $N\times (P^{2}\times C)$, where $(H, W)$ is the resolution of the original image, $C$ is the number of channels, $(P, P)$ is the resolution of each image patch, and $N=\frac{HW}{P^{2}}$ is the resulting number of patches.

![image](https://user-images.githubusercontent.com/37235023/216051924-94cb2336-d1a9-4329-b4d5-c2ea3cb3d6a7.png)
 - **Eq. 1**: The [Transformer](https://sh-tsang.medium.com/review-attention-is-all-you-need-transformer-96c787ecdec1) uses constant latent vector size D through all of its layers, so the patches are flattened and map to D dimensions with a trainable linear projection $E$. The output of this projection as the patch embeddings.
 - Similar to [BERT](https://sh-tsang.medium.com/59b1684882db)’s [class] token, a learnable embedding $x_{class}$ is prepended to the sequence of embedded patches ( $z_{00}=x_{class}$ ).
 - **Eq. 4**: The state at the output of the [Transformer](https://sh-tsang.medium.com/review-attention-is-all-you-need-transformer-96c787ecdec1) encoder ( $z_{L}^{0}$ ) serves as the image representation $y$.
 - Both during pre-training and fine-tuning, a classification head is attached to $z_{L}^{0}$. The classification head is implemented by a MLP with one hidden layer at pre-training time and by a single linear layer at fine-tuning time.
 - Position embeddings are added to the patch embeddings to retain positional information. Standard learnable 1D position embeddings is used.
 - **Eq. 2, 3**: The [Transformer](https://sh-tsang.medium.com/review-attention-is-all-you-need-transformer-96c787ecdec1) encoder consists of alternating layers of multiheaded self-attention (MSA) and MLP blocks.
 - Layernorm (LN) is applied before every block, and residual connections after every block. The MLP contains two layers with a [GELU](https://sh-tsang.medium.com/review-gaussian-error-linear-units-gelus-d4d7347d1e11) non-linearity.

Details of Vision Transformer model variants.
![image](https://user-images.githubusercontent.com/37235023/216055819-5b6558de-cd3c-435d-bb14-27b1f3696bb8.png)
 - The “Base” and “Large” models are directly adopted from [BERT](https://sh-tsang.medium.com/59b1684882db) and the larger “Huge” model is added.
 - ViT-L/16 means the “Large” variant with 16×16 input patch size. Note that the [Transformer](https://sh-tsang.medium.com/review-attention-is-all-you-need-transformer-96c787ecdec1)’s sequence length is inversely proportional to the square of the patch size, and models with smaller patch size are computationally more expensive.

**Comment by NorbertZheng** [comment]
## Some Training Details
 - Typically, ViT is pre-trained on large datasets, and fine-tuned to (smaller) downstream tasks. For this, the pre-trained prediction head is removed and a zero-initialized $D\times K$ feedforward layer is attached, where $K$ is the number of downstream classes.
 - It is often beneficial to fine-tune at a higher resolution than pre-training, as in [FixRes](https://sh-tsang.medium.com/f533434538e9). When feeding images of higher resolution, the patch size is kept the same, which results in **a larger effective sequence length**.
 - The Vision Transformer can handle arbitrary sequence lengths (up to memory constraints), however, the pre-trained position embeddings may no longer be meaningful. Therefore 2D interpolation of the pre-trained position embeddings is performed, according to their location in the original image. (This resolution adjustment and patch extraction are the only points at which an inductive bias about the 2D structure of the images is manually injected into the Vision Transformer.)

**Comment by NorbertZheng** [comment]
## Experimental Results
### SOTA Comparison
![image](https://user-images.githubusercontent.com/37235023/216226294-3a45f87d-8326-4a42-8e41-f6257e70e845.png)
Comparison with state of the art on popular image classification benchmarks. The number of TPUv3-core-days is the number of TPU v3 cores (2 per chip) used for training multiplied by the training time in days.

The smaller ViT-L/16 model pre-trained on JFT-300M outperforms [BiT](https://sh-tsang.medium.com/cb4bf8ed9732)-L on all tasks, while requiring substantially fewer computational resources to train.
 - The larger model, ViT-H/14, further improves the performance, especially on the more challenging datasets.
 - ViT models still took substantially less compute to pre-train.

![image](https://user-images.githubusercontent.com/37235023/216226746-4a28ddcc-9b29-4cf4-aa32-ba010449cfec.png)
Breakdown of VTAB performance in Natural, Specialized, and Structured task groups.
 - [BiT](https://sh-tsang.medium.com/cb4bf8ed9732), VIVI — a [ResNet](https://towardsdatascience.com/review-resnet-winner-of-ilsvrc-2015-image-classification-localization-detection-e39402bfa5d8?source=post_page---------------------------) co-trained on ImageNet and YouTube, and S4L — supervised plus semi-supervised learning on ImageNet.

ViT-H/14 outperforms [BiT](https://sh-tsang.medium.com/cb4bf8ed9732)-R152×4, and other methods, on the Natural and Structured tasks. On the Specialized the performance of the top two models is similar.

**Comment by NorbertZheng** [comment]
### Pretraining Data Requirement
![image](https://user-images.githubusercontent.com/37235023/216337516-dba8df33-b06a-4590-9e2a-edacf66efe6e.png)
Left: Transfer to ImageNet, Right: Linear few-shot evaluation on ImageNet versus pre-training size (ViT-b is ViT-B with all hidden dimensions halved.)

 - **Left**: When pre-trained on the smallest dataset, ImageNet, ViT-Large models underperform compared to ViT-Base models, despite (moderate) regularization.
 - With ImageNet-21k pre-training, their performances are similar.
 - Only with JFT-300M, we can see the full benefit of larger models.
 - The [BiT](https://sh-tsang.medium.com/cb4bf8ed9732) CNNs outperform ViT on ImageNet, but with the larger datasets, ViT overtakes.
 - **Right**: Random subsets of 9M, 30M, and 90M as well as the full JFT-300M dataset for pretraining.
 - ViT-B/32 is slightly faster than [ResNet](https://towardsdatascience.com/review-resnet-winner-of-ilsvrc-2015-image-classification-localization-detection-e39402bfa5d8?source=post_page---------------------------)50; it performs much worse on the 9M subset, but better on 90M+ subsets.
 - The same is true for [ResNet](https://towardsdatascience.com/review-resnet-winner-of-ilsvrc-2015-image-classification-localization-detection-e39402bfa5d8?source=post_page---------------------------)152×2 and ViT-L/16.

This result reinforces the intuition that the convolutional inductive bias is useful for smaller datasets, but for larger ones, learning the relevant patterns directly from data is sufficient, even beneficial.

**Comment by NorbertZheng** [comment]
## Scaling Study
![image](https://user-images.githubusercontent.com/37235023/216338334-743b6a41-fdbf-404c-9ac8-4e0a9663f9c1.png)
Performance versus pre-training compute for different architectures (The number of pretraining epochs is from 7 to 14, so that data size does not bottleneck the models’ performances).
 - Vision Transformers dominate [ResNet](https://towardsdatascience.com/review-resnet-winner-of-ilsvrc-2015-image-classification-localization-detection-e39402bfa5d8?source=post_page---------------------------)s on the performance/compute trade-off. ViT uses approximately 2-4× less compute to attain the same performance (average over 5 datasets).
 - Hybrids slightly outperform ViT at small computational budgets, but the difference vanishes for larger models. This result is somewhat surprising, since one might expect convolutional local feature processing to assist ViT at any size.
 - Vision Transformers appear not to saturate within the range tried.

**Comment by NorbertZheng** [comment]
### Inspecting Vision Transformer
![image](https://user-images.githubusercontent.com/37235023/216338746-8762a8b3-069b-4b81-95a2-1512aee341f2.png)
Left: Filters of the initial linear embedding of RGB values of ViT-L/32. Center: Similarity of position embeddings of ViT-L/32. Right: Size of attended area by head and network depth.
 - **Left**: The components resemble plausible basis functions for a low-dimensional representation of the fine structure within each patch.
 - **Center**: The model learns to encode distance within the image in the similarity of position embeddings, i.e. closer patches tend to have more similar position embeddings.
 - **Right**: The average distance in image space across which information is integrated, based on the attention weights. This “attention distance” is analogous to receptive field size in CNNs.
 - Some heads attend to most of the image already in the lowest layers, showing that the ability to integrate information globally is indeed used by the model.

![image](https://user-images.githubusercontent.com/37235023/216338994-f63c8a45-1f59-419d-960f-124714c648e0.png)
Representative examples of attention from the output token to the input space.

![image](https://user-images.githubusercontent.com/37235023/216339126-a46aa052-38f4-4515-a9c5-2a1ef71b628e.png)
Further example attention maps.

The model attends to image regions that are **semantically relevant** for classification.

**Comment by NorbertZheng** [comment]
### Self-Supervision
A preliminary exploration is studied on masked patch prediction for self-supervision, mimicking the masked language modeling task used in [BERT](https://sh-tsang.medium.com/59b1684882db).

With self-supervised pre-training, the smaller ViT-B/16 model achieves 79.9% accuracy on ImageNet, a significant improvement of 2% to training from scratch, but still 4% behind supervised pre-training.

**Comment by NorbertZheng** [comment]
## Reference
 - [2021 ICLR] [Vision Transformer, ViT]
[An Image is Worth 16x16 Words Transformers for Image Recognition at Scale](https://arxiv.org/abs/2010.11929).

**Comment by NorbertZheng** [comment]
## Image Classification
1989–2018 … 2019: [[ResNet-38](https://medium.com/@sh.tsang/resnet-38-wider-or-deeper-resnet-image-classification-semantic-segmentation-f297f2f73437)] [[AmoebaNet](https://medium.com/@sh.tsang/reading-amoebanet-regularized-evolution-for-image-classifier-architecture-search-image-278f5c077a4a)] [[ESPNetv2](https://medium.com/@sh.tsang/reading-espnetv2-a-light-weight-power-efficient-and-general-purpose-convolutional-neural-a57b5fbfeb84)] [[MnasNet](https://sh-tsang.medium.com/reading-mnasnet-platform-aware-neural-architecture-search-for-mobile-image-classification-b042aaef66f7)] [[Single-Path NAS](https://medium.com/carre4/paper-single-path-nas-designing-hardware-efficient-convnets-in-less-than-4-hours-image-classifi-81cff87f223e)] [[DARTS](https://sh-tsang.medium.com/paper-darts-differentiable-architecture-search-image-classification-dda80703b81a)] [[ProxylessNAS](https://sh-tsang.medium.com/paper-proxylessnas-direct-neural-architecture-search-on-target-task-image-classification-73c35ebd8aed)] [[MobileNetV3](https://sh-tsang.medium.com/paper-mobilenetv3-searching-for-mobilenetv3-image-classification-5072d4d8703c)] [[FBNet](https://sh-tsang.medium.com/paper-fbnet-hardware-aware-efficient-convnet-design-via-differentiable-neural-architecture-f95a20c1cd4a)] [[ShakeDrop](https://sh-tsang.medium.com/paper-shakedrop-shakedrop-regularization-for-deep-residual-learning-image-classification-d2b83d0ae3de)] [[CutMix](https://sh-tsang.medium.com/paper-cutmix-regularization-strategy-to-train-strong-classifiers-with-localizable-features-5527e29c4890)] [[MixConv](https://sh-tsang.medium.com/mixconv-mixed-depthwise-convolutional-kernels-image-classification-9d0ca8bf4093)] [[EfficientNet](https://sh-tsang.medium.com/efficientnet-rethinking-model-scaling-for-convolutional-neural-networks-image-classification-ef67b0f14a4d)] [[ABN](https://sh-tsang.medium.com/review-abn-attention-branch-network-for-visual-explanation-image-classification-30fd164960ee)] [[SKNet](https://sh-tsang.medium.com/review-sknet-selective-kernel-networks-image-classification-63ebbad7d78f)] [[CB Loss](https://sh-tsang.medium.com/review-cb-loss-class-balanced-loss-based-on-effective-number-of-samples-image-classification-3056a1a1a001)] [[AutoAugment, AA](https://sh-tsang.medium.com/review-autoaugment-learning-augmentation-strategies-from-data-image-classification-af27dea9a839)] [[BagNet](https://sh-tsang.medium.com/review-bagnet-approximating-cnns-with-bag-of-local-features-models-works-surprisingly-well-on-125f4295c433)] [[Stylized-ImageNet](https://sh-tsang.medium.com/review-stylized-imagenet-imagenet-trained-cnns-are-biased-towards-texture-increasing-shape-be4986004de)] [[FixRes](https://sh-tsang.medium.com/f533434538e9)] [[Ramachandran’s NeurIPS’19](https://sh-tsang.medium.com/review-stand-alone-self-attention-in-vision-models-55e4dbeb064c)] [[SE-WRN](https://sh-tsang.medium.com/review-se-wrn-squeeze-and-excitation-wide-residual-networks-in-image-classification-a748be573d0f)] [[SGELU](https://sh-tsang.medium.com/review-symmetrical-gaussian-error-linear-units-sgelus-e6a07a2eeead)]
2020: [[Random Erasing (RE)](https://sh-tsang.medium.com/random-erasing-re-random-erasing-data-augmentation-image-classification-37f11627b38)] [[SAOL](https://sh-tsang.medium.com/review-saol-spatially-attentive-output-layer-image-classification-weakly-supervised-object-742e2eeb3226)] [[AdderNet](https://sh-tsang.medium.com/review-addernet-do-we-really-need-multiplications-in-deep-learning-image-classification-b72851ddb255)] [[FixEfficientNet](https://sh-tsang.medium.com/c49a5e610c15)]
2021: [[Learned Resizer](https://sh-tsang.medium.com/review-learning-to-resize-images-for-computer-vision-tasks-image-classification-image-quality-c05cbe284bc6)] [[Vision Transformer, ViT](https://sh-tsang.medium.com/review-vision-transformer-vit-406568603de0)]
